### PR TITLE
fix(adr-063): close Phase 4's .abicheck.yml-only dry-run parity gap

### PR DESCRIPTION
**Pull request by napetrov**
## Summary

Closes the second half of ADR-063 Phase 4's dry-run/execution parity gap: `dump --dry-run`/`compare --dry-run`/most of `scan`'s pre-flight paths now catch a root-target scope declared only in an auto-discovered `.abicheck.yml` combined with a pre-captured Bazel `aquery`/`cquery` jsonproto, the same way the real run already did for an explicit `--build-target`.

## Motivation

`docs/contribute/known-gaps.md` and the ADR-063 implementation plan named this as an open gap after Phase 4's first slice: `AnalysisPlanner` (and `scan`'s equivalent pre-flight checks) could reject the explicit `--build-target` + pre-captured Bazel jsonproto combination, but had no way to see a root-target scope that came only from `.abicheck.yml`'s `build.targets:` — so a dry-run preview could report success for a request the real run then rejected with `PlanningError`/exit 64.

## Changes

- `workflows/plan.py`: `bazel_target_scoping_failure`/`scan_bazel_scoping_failure` gained optional `sources`/`build_config` parameters. When the request's own `build_targets` is empty, the check now falls back to an explicit `build_config` or an auto-discovered `.abicheck.yml` at `sources`, mirroring `embed_build_source`'s own `targets=list(build_targets) if build_targets else cfg.targets` precedence — never running the P0.3 compile-context fold itself, just a pure config read.
- Wired into `dump`/`compare` via `_check_bazel_target_scoping` (`AnalysisPlanner`) — closes their `--dry-run` parity for free, since dry-run rendering already resolves through the same chokepoint.
- Wired into 3 of `scan`'s 4 pre-flight call sites: `scan_engine.run_scan_core`, and both of `cli_scan.py`'s direct checks (`scan_cmd`'s single-binary pre-flight, `_run_artifact_set`'s own).
- `service_scan.run_scan_set` was deliberately **not** widened: that file sits exactly at the AI-readiness 2000-line hard cap, and the widened call doesn't fit `ruff format`'s line budget. Documented as an explicit, narrow residual (a direct `run_scan_set(ScanRequest(...))` typed-API call with no CLI in front of it; `scan --artifact-set`'s own CLI path is unaffected since its pre-flight runs first) rather than forced in under a file-size constraint.
- Updated `docs/contribute/adr/063-one-semantic-pipeline.md` (Phase 4 "Second slice"), `docs/contribute/plans/one-semantic-pipeline.md`, and `docs/contribute/known-gaps.md` with the full accounting; fixed a stale `adr/index.md` summary line that still said "Phases 4-10 not yet" after Phase 4's first slice had already landed.
- Added a `changelog.d/` fragment.

## Bug-fix test contract

- Bug class: Dry-run/real-execution parity gap — a pre-flight check reads only an explicit request-level flag and misses an equivalent value sourced from project config (`.abicheck.yml`), so a preview can report success for a request the real run then rejects.
- Publicly observable failure: `dump --dry-run`/`compare --dry-run` (and 3 of 4 `scan` pre-flight paths) previewed success for a root-target scope declared only in an auto-discovered `.abicheck.yml`'s `build.targets:` combined with a pre-captured Bazel jsonproto `--build-info`, when the real run raised `PlanningError` (exit 64).
- Regression test fails on base: Yes — e.g. `tests/test_bazel_root_targets.py::test_dot_abicheck_yml_build_targets_dry_run_parity` and `tests/test_analysis_plan.py::TestBazelBuildTargetScoping::test_dump_config_sourced_target_scope_raises_planning_error` fail against the pre-fix code (neither `AnalysisPlanner` nor the dry-run path saw the config-sourced scope at all).
- Negative control: `test_no_abicheck_yml_present_is_unaffected`/pre-existing `test_no_build_targets_is_unaffected` assert an unscoped request or one with no `.abicheck.yml` present is not rejected.
- Public-surface test: `test_dot_abicheck_yml_build_targets_dry_run_parity` drives the real `dump --dry-run` CLI via `CliRunner`, asserting exit code 64 and message content — the user-facing surface, not internals.
- Axes covered: `dump` (`AnalysisPlanner`, dry-run and real), `compare` (either side), `scan` (3 of 4 pre-flight sites: `run_scan_core` real+dry-run, single-binary CLI, `--artifact-set` CLI); explicit `build_targets` still wins over a present config (precedence + message wording); the pre-existing `depth=binary` exemption still holds with a config-sourced scope; a malformed `.abicheck.yml` degrades to "no config found" rather than raising a second error; no-config-present is unaffected.
- General invariant: the scoping pre-flight check must see the same effective target list real execution computes (explicit `build_targets` if given, else `.abicheck.yml`'s `build.targets:`) regardless of front end or preview-vs-real-run mode — verified across independently-chosen call sites (`dump`, `compare`, three `scan` entry points) and edge cases (precedence, exemptions, malformed/absent config), not just the one originally-named scenario.

## Checklist

- [x] Tests added / updated for new behaviour
- [x] Changelog fragment added (`scriv create`)
- [x] Docs updated (ADR-063, implementation plan, known-gaps.md, `adr/index.md`)
- [x] `mypy`/`ruff check`/`ruff format --check` clean on touched files
- [x] AI-readiness gate (`scripts/check_ai_readiness.py`) clean (0 errors)

---
_Generated by [Claude Code](https://claude.ai/code/session_01XUdAZRSGwX1DJNK9HSQz3b)_

## Summary by CodeRabbit

* **Bug Fixes**
  * Dry runs for `dump`, `compare`, and `scan` now correctly reject disallowed Bazel root-target scopes configured in `.abicheck.yml`.
  * Scan validation now consistently honors explicit targets, discovered configuration, effective depth, and header-related exemptions.
  * Improved validation and error handling for single-binary and artifact-set scans.

* **Documentation**
  * Updated architecture records, plans, known gaps, and release documentation to reflect validation parity improvements.

* **Tests**
  * Added regression and end-to-end coverage for configuration discovery, dry-run behavior, target scoping, and scan exemptions.

### CODE REVIEW

File: abicheck/cli_scan.py [hunk 1]
```diff
@@ -1793,7 +1799,11 @@ def scan_cmd(
     from .workflows.plan import bazel_target_scoping_failure
 
     if _bf := bazel_target_scoping_failure(
-        "candidate", effective_build_info, build_targets
+        "candidate",
+        effective_build_info,
+        build_targets,
+        sources=sources,
+        build_config=build_config,
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/cli_scan.py:1813]: Confirmed and fixed in 78d2b89a: routed `scan_cmd`'s single-binary pre-flight through `scan_bazel_scoping_failure` instead of the bare check, so it now carries the headers/collect-mode exemption. Pinned with a regression test.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: abicheck/workflows/plan.py [hunk 2]
```diff
@@ -213,8 +213,71 @@ def _replace_lang_frontend(plan: SidePlan, lang: str, frontend: str) -> SidePlan
...
+    """
+    if build_config is not None:
+        cfg_path: Path | None = build_config
+    else:
+        if sources is None:
+            return ()
+        from ..buildsource.pack_shape import is_pack_dir
+
+        if is_pack_dir(sources):
+            return ()
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/workflows/plan.py]: Confirmed and fixed in 09de29fb: switched to `is_any_pack_dir()`, which recognizes both pack shapes. Pinned with a regression test using a Flow-2 `abicheck_inputs` pack with a bundled `.abicheck.yml`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: abicheck/cli_scan.py [hunk 3]
```diff
@@ -756,7 +756,13 @@ def _run_artifact_set(
     from .workflows.plan import bazel_target_scoping_failure
 
     if (depth or "").lower() != "binary" and (
-        _bf := bazel_target_scoping_failure("candidate", build_info, build_targets)
+        _bf := bazel_target_scoping_failure(
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/cli_scan.py]: Confirmed and fixed in 528111c3: `_run_artifact_set` has no per-member resolved `collect_mode` at this point, so I added `artifact_set_bazel_scoping_failure` mirroring `_check_bazel_target_scoping`'s own raw-depth reasoning rather than reusing `scan_bazel_scoping_failure`'s signature. Pinned with a regression test.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: abicheck/workflows/plan.py [hunk 4]
```diff
@@ -213,8 +214,81 @@ def _replace_lang_frontend(plan: SidePlan, lang: str, frontend: str) -> SidePlan
...
+
+    if build_config is not None:
+        cfg_path: Path | None = build_config
+    else:
+        if sources is None:
+            return ()
+        from ..buildsource.inputs_pack import is_any_pack_dir
+
+        if is_any_pack_dir(sources):
+            return ()
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/workflows/plan.py]: Verified against `buildsource/l2_seed.py` directly and confirmed — `_l2_seed_config` does call `discover_build_config` on the original pack path before pack recognition nulls it, and that seed runs whenever headers are present (`seed_includes_and_fold_compile_context`'s own early-return gate), independent of collect_mode. Fixed in 677fdd2f by threading a `headers_present` signal through `bazel_target_scoping_failure` so the pack skip only applies when headers are absent. Pinned with a regression test (pack + real headers + bundled config now correctly raises).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: abicheck/workflows/plan.py [hunk 5]
```diff
@@ -378,7 +484,44 @@ def _check_bazel_target_scoping(side: SidePlan) -> PlanningFailure | None:
...
+    via :func:`_depth_implied_collect_mode`; only ``"binary"`` additionally
+    clears headers) rather than ``scan_bazel_scoping_failure``'s, which needs a
+    real, resolved ``EvidenceDepth``/``collect_mode`` pair this call site
+    doesn't have yet.
+    """
+    is_binary = (depth or "").lower() == "binary"
+    effective_headers_present = False if is_binary else headers_present
+    off = depth is not None and _depth_implied_collect_mode(depth) == "off"
+    if off and not effective_headers_present:
+        return None
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/workflows/plan.py]: Confirmed and fixed in 677fdd2f, though narrower than "pass the actual resolved level" — that would need duplicating real risk-scoring resolution at this pre-flight stage, which `AnalysisPlan`'s own design (this module's docstring) deliberately excludes. Instead, an unset `--depth` now withholds only the newly-added config-sourced fallback (can't safely rule "off" in or out), while an *explicit* `--build-target` is still checked exactly as before — preserving the two pre-existing tests that pin that established behavior. Pinned with a regression test for the config-sourced case.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: abicheck/workflows/plan.py [hunk 6]
```diff
@@ -255,8 +343,12 @@ def bazel_target_scoping_failure(
     fmt = sniff_build_info_format(build_info)
     if fmt not in ("bazel_aquery", "bazel_cquery"):
         return None
+    source_note = "" if build_targets else " (from an auto-discovered .abicheck.yml)"
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/workflows/plan.py]: Applied your suggested fix as-is in 3db96a3b, plus a direct unit test against `bazel_target_scoping_failure` (dump/compare have no request-level seam for an explicit `build_config` yet, so `AnalysisPlanner` itself can't exercise this branch).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: changelog.d/20260901_034418_noreply_adr_063_phase_completion_wcvpef.md [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+  `--build-target`.** A pre-captured Bazel `aquery`/`cquery` `--build-info`
+  jsonproto combined with a root-target scope silently ran unscoped
+  (collecting every action/target in the captured graph); a prior fix
+  raised a clear usage error (exit 64) for an explicit `--build-target`
+  flag, but a scope declared only in a discovered `.abicheck.yml`'s
+  `build.targets:` still previewed success under `--dry-run` for a request
+  the real run then rejected. `dump`/`compare`/`scan`'s pre-flight check
+  now auto-discovers (or, for `scan`, honors an explicit `--config`
+  override for) that config the same way real execution does, closing the
+  gap for every shape of the request — CLI and typed Python API alike.
```
[the remark this answers is not in the document]
Comment (napetrov) [reply changelog.d/20260901_034418_noreply_adr_063_phase_completion_wcvpef.md]: Fixed in 3db96a3b: narrowed both the changelog fragment and `adr/index.md`'s Phase 4 summary to name `service_scan.run_scan_set`'s direct typed-API call as the one remaining, deliberate boundary instead of the unqualified "alike" claim.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: abicheck/workflows/plan.py [hunk 8]
```diff
@@ -378,7 +513,69 @@ def _check_bazel_target_scoping(side: SidePlan) -> PlanningFailure | None:
...
+    outright, and the shape every existing test/caller since this phase's
+    first slice already exercises. What changes for an unset *depth* is
+    narrower: the *config-sourced* (``.abicheck.yml``-only, no explicit
+    ``build_targets``) fallback is withheld by not forwarding
+    *sources*/*build_config* at all -- trusting an auto-discovered scope
+    this check cannot rule "reachable" or not for is the part that's newly
+    added and newly risky; an explicit scope the caller named is not.
+    """
+    if depth is None:
+        return bazel_target_scoping_failure("candidate", build_info, build_targets)
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/workflows/plan.py]: Confirmed and fixed in 9dac2e58. Rather than approximating the resolved level at this pre-flight (which `AnalysisPlan`'s own design deliberately excludes), `_run_artifact_set` now resolves the *real* `eff_depth`/`collect_mode` via `service_scan._resolve_member_scan_level` -- the same primitive `estimate_artifact_set`'s own `--dry-run` cost totals already use -- before calling the existing, already-shared `scan_bazel_scoping_failure`. This replaces the bespoke `workflows.plan.artifact_set_bazel_scoping_failure` (deleted) that approximated an unset `--depth`, which two earlier rounds on this same thread had already shown gets both directions wrong (always-active over-rejects a real no-op; always-exempt under-rejects exactly the case you're flagging here). Pinned with two regression tests using a real low-risk vs. high-risk `--changed-path` seed: `test_scan_cli_artifact_set_unset_depth_low_risk_seed_config_scope_is_unaffected` (exit 0) and `test_scan_cli_artifact_set_high_risk_seed_config_scope_still_rejects` (exit 64). Full account in `docs/contribute/known-gaps.md`'s "thirteenth review round" entry.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: abicheck/cli_scan.py [hunk 9]
```diff
@@ -743,20 +743,34 @@ def _run_artifact_set(
...
+    from .workflows.plan import scan_bazel_scoping_failure
+
+    # ADR-063 Phase 4 (Codex review): checked before discovery, via the real
+    # resolved eff_depth/collect_mode (same primitive estimate_artifact_set's
+    # --dry-run totals use), not a raw --depth approximation (both
+    # over/under-rejected in earlier rounds).
+    changed, changed_src, seeded = _resolve_changed_seed(
+        changed_paths_opt, since, sources
+    )
+    _, _, _, _, _, _, eff_depth, collect_mode = _resolve_member_scan_level(
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/cli_scan.py]: Confirmed and fixed in 1274bebc. Reproduced the exact regression against `TestArtifactSetMalformedRiskRules::test_malformed_risk_rules_yaml_is_usage_error` and its `--dry-run` sibling in `test_scan_artifact_set_coverage.py` before fixing -- both now pass. Wrapped the new `_resolve_member_scan_level` call in its own `try`/`except ValueError`, translating to `click.UsageError` the same way the two pre-existing blocks in this function already do. Full account in `docs/contribute/known-gaps.md`'s "fourteenth review round" entry.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: abicheck/workflows/plan.py [hunk 10]
```diff
@@ -274,26 +392,44 @@ def scan_bazel_scoping_failure(
     collect_mode: str,
     build_info: Path | None,
     build_targets: tuple[str, ...],
+    sources: Path | None = None,
+    build_config: Path | None = None,
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/workflows/plan.py:396]: This is the one gap this PR leaves open deliberately, not missed: `service_scan.py` sits exactly at the AI-readiness 2000-line hard cap, and forwarding `sources=`/`build_config=` at this call site doesn't fit `ruff format`'s column budget on one line without pushing the file over. It's named explicitly as a known, tracked residual in `docs/contribute/known-gaps.md`, `docs/contribute/plans/one-semantic-pipeline.md`'s Phase 4 status, `docs/contribute/adr/index.md`, and this PR's own changelog fragment -- a direct `run_scan_set(ScanRequest(...))` typed-API call with no CLI in front of it still rejects a configuration-only scope later, during real embedding, rather than at this pre-flight. `scan --artifact-set`'s own CLI path is unaffected: `cli_scan._run_artifact_set`'s pre-flight already forwards both values ahead of this call, so it catches the mismatch first regardless. A future pass splitting `service_scan.py` under its own file-size budget would remove this constraint; not attempted reactively here, per this ADR's own governing "acknowledged gap over risky reactive patch" convention.

---
_Generated by [Claude Code](https://claude.ai/code)_
